feat: [#601] Add filter collapse/expansion mechanism

wujuu opened the pull request:
Closes #601

wujuu commented on ui/apps/ui/src/app/components/filters/filter-multiselect/first-n-values.component.ts:
```diff
@@ -45,11 +45,27 @@ export class FirstNValuesComponent implements OnChanges {
         search(this.query, this.allEntities)
       ).slice(0, this.displayMax);
     }
+
+    if (changes['allEntities'].currentValue) {
+      // Mapping so we don't get TS errors
+      const currentValues: { isSelected: boolean }[] =
+        changes['allEntities'].currentValue;
+
+      // Workaround for Error: NG0100: ExpressionChangedAfterItHasBeenCheckedError
+      Promise.resolve().then(() => {
+        this.hasSelectedFiltersOutput.emit(
+          currentValues.some(({ isSelected }) => isSelected)
+        );
+      });
+    }
```
I'd like some advice on a better way of handling this perhaps?

Anok666 replied:
**import { ChangeDetectionStrategy } from '@angular/core';**
@Component({
  selector: 'ess-first-n-values',
  **changeDetection: ChangeDetectionStrategy.OnPush**
  template: ` <div class="filter__viewport">
    <ess-checkboxes-tree
      [data]="_allEntities"
      (checkboxesChange)="toggleActive.emit($event)"
    ></ess-checkboxes-tree>
  </div>`,
  styles: [
    `
      .filter__viewport {
        max-height: 290px;
        overflow: auto;
      }
    `,
  ],
})

wujuu replied:
I've changed the logic of letting the filters know that there aren't any active. Now I've added some logic to the _customRoute observables in the multiselect component. This should be clearer

wujuu commented on ui/apps/ui/src/app/components/filters/filter-multiselect/show-all.component.ts:
```diff
@@ -63,6 +66,19 @@ export class ShowAllComponent implements OnChanges {
       );
       this._latestChunk = 0;
     }
+
+    if (changes['allEntities'].currentValue) {
+      // Mapping so we don't get TS errors
+      const currentValues: { isSelected: boolean }[] =
+        changes['allEntities'].currentValue;
+
+      // Workaround for Error: NG0100: ExpressionChangedAfterItHasBeenCheckedError
+      Promise.resolve().then(() => {
+        this.hasSelectedFiltersOutput.emit(
+          currentValues.some(({ isSelected }) => isSelected)
+        );
+      });
+    }
```
I'd like some advice on a better way of handling this perhaps?

agpul reviewed (changes requested):
Collapse on year range doesn't work
![image](https://github.com/cyfronet-fid/eosc-search-service/assets/43000481/8ecda2fe-f85e-43a3-b335-12c09d5239e8)

wujuu commented:
I've fixed the date range and range components collapse logic (@agpul). Also adjusted the default collapse logic according to the screenshots (@aleksandrahojszyk )

Marcelinna commented:
now the "clear all" button is visible only when I select a filter, it is intentional?
pr
![image](https://github.com/cyfronet-fid/eosc-search-service/assets/59872096/c20fac40-a276-42b7-aeef-a13289b64e92)
production
![image](https://github.com/cyfronet-fid/eosc-search-service/assets/59872096/c4f4e5d6-f387-4951-905f-1c966e307dfb)

wujuu commented:
> now the "clear all" button is visible only when I select a filter, it is intentional?
> 
> pr
> 
> ![image](https://github.com/cyfronet-fid/eosc-search-service/assets/59872096/c20fac40-a276-42b7-aeef-a13289b64e92)
> 
> production
> 
> ![image](https://github.com/cyfronet-fid/eosc-search-service/assets/59872096/c4f4e5d6-f387-4951-905f-1c966e307dfb)
> 
> 

Yes, as requested by @aleksandrahojszyk

michal-szostak commented on ui/apps/ui/src/app/collections/repositories/types.ts:
```diff
@@ -98,6 +98,7 @@ export interface IFilterConfig {
   type: 'multiselect' | 'select' | 'date' | 'tag' | 'range';
 
   onFacetsFetch?: (bucketValues: IFacetBucket[]) => IFilterNode[]; // !!! only for multiselect !!!
+  defaultCollapsed?: boolean;
```
I think it introduces issues later to allow booleans to have 3 values (`true`, `false` and `undefined`).
It would be better to restructure code to accept only boolean.

But this is a minor. It can stay "as is"

wujuu replied:
done

michal-szostak commented on ui/apps/ui/src/app/components/filters/filter-range/filter-range.component.ts:
```diff
@@ -130,4 +140,8 @@ export class FilterRangeComponent implements OnInit {
     return toRangeTimeFormat(seconds);
   }
+
+  _isExpandedChanged(newExpanded: boolean) {
```
I'd advice against public methods being prefixed with `_`

wujuu replied:
done

michal-szostak reviewed (approved):
A couple of minor thing which could be fixed, but overall it's good.

wujuu commented on ui/apps/ui/src/app/components/filters/filters-layouts/filter-label.component.ts:
```diff
@@ -37,4 +82,9 @@ export class FilterLabelComponent {
     });
   }
+
+  _toggleExpanded() {
```
done
